meta-client/src/main/java/org/apache/helix/metaclient/impl/zk/util/ZkMetaClientUtil.java

```diff
@@ -358,4 +358,25 @@ public static MetaClientException.ReturnCode translateZooKeeperCodeToMetaClientC
         return MetaClientException.ReturnCode.DB_USER_ERROR;
     }
   }
+
+  public static String getZkParentPath(String path) {
+    if (path.equals("/")) {
+      return null;
+    }
+
+    int idx = path.lastIndexOf('/');
+    return idx == 0 ? "/" : path.substring(0, idx);
+  }
+
+  // Splits a path into the paths for each node along the way.
+  // /a/b/c --> /a/b/c, /a/b, /a
+  public static List<String> separateIntoUniqueNodePaths(String path) {
```

Comment:
I would suggest the algorithm is like this:
List\<String\> nodePaths = new ArrayList<>();
String[] subPath = path.split("/");
StringBuilder tempPath = new StringBuilder();
for (int i = subPath.length - 1; i >= 1; i--) {
    tempPath.insert(0, subPath[I]);
    tempPath.insert(0, "/");
    nodePaths.add(tempPath);
}

So this guarantee the list ordered by top parent path and you don't have to search for index over and over again, which O(n^2) as n chars for paths.

Reply:
I think this would take /a/b/c and return:
/c, /b/c,and /a/b/c
I've taken this algo and made a small adjustment to it